### PR TITLE
Relationship capabilities and bugfixes

### LINKED ISSUE #1187: Add equivalent of AttrCapabilities for relationships
## Is your feature request related to a problem?
This design is based on the community request at https://gitter.im/json-api-dotnet-core/Lobby?at=63199f777ccf6b6d45c52943.

## Describe the solution you'd like
Currently, developers can control [which operations are allowed declaratively per JSON:API _attribute_](https://www.jsonapi.net/usage/resources/attributes.html#capabilities) in the following way:

```c#
public class Customer : Identifiable<long>
{
    [Attr(Capabilities = AttrCapabilities.AllowFilter | AttrCapabilities.AllowSort)]
    public string Name { get; set; }
}
```

If no capabilities are set in `[Attr]`, what's set in `IJsonApiOptions.DefaultAttrCapabilities` (which defaults to `AttrCapabilities.All`) is used.

The proposal here is to add a similar feature for relationships. It would enable the next example:

```c#
public class Customer : Identifiable<long>
{
    [Attr(Capabilities = AttrCapabilities.AllowFilter | AttrCapabilities.AllowSort)]
    public string Name { get; set; }

    [HasOne(Capabilities = HasOneCapabilities.AllowView | HasOneCapabilities.AllowSet)]
    public Address? Address { get; set; }

    [HasMany(Capabilities = HasManyCapabilities.AllowView | HasManyCapabilities.AllowAdd)]
    public ISet<Order> Orders = new HashSet<Order>();
}
```

The effective relationship capabilities would likewise be determined by falling back to what's set on `IJsonApiOptions`, defaulting to `All`. When an incoming request violates these constraints, an appropriate HTTP error is returned for one of the violations (the order is undefined and we won't aggregate all violations; I'm assuming that's how `AttrCapabilities` works today). Similarly to how `AttrCapabilities` is implemented, custom logic in resource definitions is not constrained by relationship capabilities. In case `AllowView` is blocked, we need to take precautions to never return the relationship in any response bodies.

Proposed capabilities to add:
```c#
/// <summary>
/// Indicates capabilities that can be performed on a <see cref="HasOneAttribute" />.
/// </summary>
[Flags]
public enum HasOneCapabilities
{
    None = 0,

    /// <summary>
    /// Whether or not GET requests can retrieve the to-one relationship. Attempts to retrieve when disabled
    /// will return an HTTP 400 response.
    /// </summary>
    AllowView = 1,

    /// <summary>
    /// Whether or not POST and PATCH requests can assign or clear the to-one relationship. Attempts to assign
    /// or clear when disabled will return an HTTP 400 response from relationship endpoints and an HTTP 422
    /// response from resource endpoints.
    /// </summary>
    AllowSet = 2,

    All = AllowView | AllowSet
}

/// <summary>
/// Indicates capabilities that can be performed on a <see cref="HasManyAttribute" />.
/// </summary>
[Flags]
public enum HasManyCapabilities
{
    None = 0,

    /// <summary>
    /// Whether or not GET requests can retrieve the to-many relationship. Attempts to retrieve when disabled
    /// will return an HTTP 400 response.
    /// </summary>
    AllowView = 1,

    /// <summary>
    /// Whether or not POST and PATCH requests can replace the to-many relationship. Attempts to replace
    /// when disabled will return an HTTP 400 response from relationship endpoints and an HTTP 422 response
    /// from resource endpoints.
    /// </summary>
    AllowSet = 2,

    /// <summary>
    /// Whether or not POST requests can add to the to-many relationship. Attempts to add when disabled
    /// will return an HTTP 400 response.
    /// </summary>
    AllowAdd = 4,

    /// <summary>
    /// Whether or not DELETE requests can remove from the to-many relationship. Attempts to remove when
    /// disabled will return an HTTP 400 response.
    /// </summary>
    AllowRemove = 8,

    All = AllowView | AllowSet | AllowAdd | AllowRemove
}
```

## Alternative solutions and open questions
1. List of capabilities
    We could split `Has[One|Many]Capabilities.AllowSet` into `.Clear` and `.Assign`, or add these while keeping `Set` (which implies both).
    - `.Clear` for a to-one relationship means assigning `null` and for a to-many relationship it means assigning an empty array.
    - `.Assign` for a to-one relationship means assigning a non-null value and for a to-many relationship it means assigning a non-empty array.

    We can't further distinguish between "assign from empty" vs "replace the existing value" for a to-one relationship, because it would require fetching the currently stored value first, which we're trying to avoid for performance reasons.

    The split could be done at a later time by adding the Clear/Assign members next to the existing Set member. On the other hand, there's no good reason to postpone it if it's considered useful to have them today. I can't think of a reason why the need for them would increase in the future.

2. HTTP status code
    Today JsonApiDotNetCore returns 400 for bad requests, but 422 when the problem is in the request body. Because setting a relationship could be done from:

    - POST /customers
    - PATCH /customers/1
    - PATCH /customers/1/relationships/orders

    adhering to the existing rule means to return 422 in the first two cases, but 400 in the last one. The alternative is to break with the rule and always return HTTP 400 on capabilities violations. I'm not sure what the right choice is here.

3. Meaning of `AllowView`
    When blocked, this could either mean hiding the relationship data (type + id), or the relationship itself (data + links). And this raises the question of how it interacts with includes from query strings. Today we have the boolean `CanInclude` property on HasOne/HasMany relationships, which determines whether `?include=` can contain the relationship. Should blocking `AllowView` implicitly mean to also block `?include=`, or should we leave this up to the user? The latter means more flexibility, but at the cost of unintentionally forgetting to set `CanInclude = false` in addition to blocking `AllowView`, resulting in the wrong outcome. And should we deprecate `CanInclude` in favor of adding `.AllowInclude` to the enumerations?

### LINKED ISSUE #1195: Avoid primary data to occur in included[]
#### DESCRIPTION
The primary resource is duplicated in the `included` array in rare corner cases, such as self-referencing relationships or cyclic relationship chains.

I don't know if this was a deliberate choice or just an unnoticed side effect during initial implementation (many years ago, before I joined the project). The behavior could be considered convenient, because clients parsing the `included` array can look up everything in the array, without falling back to the primary data.

We've changed the logic to build the `include` array multiple times. But before doing so, we added tests to capture the existing behavior. Therefore it's still working this way today.

The JSON:API specification used to be agnostic about this, but with its [recent rewording](https://github.com/json-api/json-api/pull/1541), this is no longer allowed:

> Every included resource object MUST be identified via a chain of relationships originating in a document’s primary data. This means that compound documents require “full linkage” and that no resource object can be included without a direct or indirect relationship to the document’s primary data.
>
> The only exception to the full linkage requirement is when relationship fields that would otherwise contain linkage data are excluded due to [sparse fieldsets](https://jsonapi.org/format/#fetching-sparse-fieldsets) requested by the client.

#### STEPS TO REPRODUCE
See cases in [IntegrationTests/QueryStrings/Includes/IncludeTests.cs](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/6b69ed3f8f396348b1665e6395acfb77ab74c4bd/test/JsonApiDotNetCoreTests/IntegrationTests/QueryStrings/Includes/IncludeTests.cs).

#### EXPECTED BEHAVIOR
The primary resource is never duplicated in the `included` array in response bodies.

#### ACTUAL BEHAVIOR
Duplication may occur.

#### VERSIONS USED
- JsonApiDotNetCore version: 5.0.3
- ASP.NET Core version: 6.0
- Entity Framework Core version: N/A
- Database provider: N/A

### LINKED ISSUE #1196: Omit included resources on empty incoming ?include=
#### DESCRIPTION

The JSON:API specification has been [updated](https://github.com/json-api/json-api/pull/1532) to require what we did for sparse fieldsets in #996 to also apply for includes.

#### STEPS TO REPRODUCE
1. Create a custom resource definition that silently includes a relationship
2. Execute GET request with query string `?include=`

#### EXPECTED BEHAVIOR
The response body contains ~either~ an empty `include` array ~or omits it entirely~.

#### ACTUAL BEHAVIOR
HTTP 400 Bad Request.

#### VERSIONS USED
- JsonApiDotNetCore version: 5.0.3
- ASP.NET Core version: 6.0
- Entity Framework Core version: N/A
- Database provider: N/A

### LINKED ISSUE #1198: IntelliSense is missing for some types originating from Annotations package
image

This is because [SauceControl.InheritDoc](https://github.com/saucecontrol/InheritDoc) (which expands `<inheritdoc />` into full doc-comments) is missing in the list of package references in `JsonApiDotNetCore.Annotations.csproj`.

### LINKED ISSUE #1199: Stepping into JADNC source does not work for Annotations package
The video below shows how stepping into `ResourceGraph` source code works, but fails when stepping into `Identifiable<int>.StringId.get`. For reasons unknown to me, the video tool failed to capture the context menu on right-click, where I selected "step into specific" to step into the `.StringId` getter.

https://user-images.githubusercontent.com/10324372/193688497-8b757cb2-8545-4ac1-afc4-dbb220dfb7e9.mp4

This is because [Microsoft.SourceLink.GitHub](https://www.nuget.org/packages/Microsoft.SourceLink.GitHub) (which enables to download sources from GitHub and step into) is missing in the list of package references in `JsonApiDotNetCore.Annotations.csproj`.

For reference, the source code from the video, in case it's hard to read:

```c#
using System.Collections;
using JsonApiDotNetCore.Configuration;
using JsonApiDotNetCore.Controllers;
using JsonApiDotNetCore.Services;
using JsonApiDotNetCoreWebApi1.Models;
using Microsoft.AspNetCore.Mvc;

namespace JsonApiDotNetCoreWebApi1.Controllers
{
    public class PeopleController:JsonApiController<Person,int>
    {
        private readonly IResourceGraph _resourceGraph;

        public PeopleController(IJsonApiOptions options, IResourceGraph resourceGraph, ILoggerFactory loggerFactory,
            IResourceService<Person, int> resourceService)
            : base(options, resourceGraph, loggerFactory, resourceService)
        {
            _resourceGraph = resourceGraph;
        }

        public override async Task<IActionResult> GetAsync(CancellationToken cancellationToken)
        {
            IActionResult result = await base.GetAsync(cancellationToken);

            _resourceGraph.FindResourceType(typeof(Person)); // <-- step into works

            var people = (IEnumerable<Person>)((OkObjectResult)result).Value!;
            
            foreach (Person person in people)
            {
                string? id = person.StringId; // <-- step into fails
            }

            return result;
        }
    }
}
```

### PR DESCRIPTION
**Pull request by bkoelman**
Adds the equivalent of `AttrCapabilities` for relationships. See #1187 for design considerations.
This PR also fixes some small bugs.

Closes #1187.
Fixes #1196.
Fixes #1195.
Fixes #1198.
Fixes #1199.

#### QUALITY CHECKLIST
- [x] Changes implemented in code
- [x] Complies with our [contributing guidelines](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/master/.github/CONTRIBUTING.md)
- [x] Adapted tests
- [x] Documentation updated

### CODE REVIEW

**Comment by bkoelman** [comment]
This PR has a dependency on #1192. If that one is merged earlier, this PR needs additional changes.

**Review by maurei — CHANGES_REQUESTED** [review]

File: docs/usage/resources/relationships.md [hunk 1]
```diff
@@ -160,7 +160,111 @@ public class TodoItem : Identifiable<int>
...
 ```
 
-## Includibility
+## Capabilities
+
+_since v5.1_
+
+Default JSON:API relationship capabilities are specified in 
+@JsonApiDotNetCore.Configuration.JsonApiOptions#JsonApiDotNetCore_Configuration_JsonApiOptions_DefaultHasOneCapabilities and 
+@JsonApiDotNetCore.Configuration.JsonApiOptions#JsonApiDotNetCore_Configuration_JsonApiOptions_DefaultHasManyCapabilities:
```
[the remark this answers is not in the document]
Comment (bkoelman) [reply docs/usage/resources/relationships.md:169]: Thanks. I'm going to try this locally first, to see if the DocFx accepts it.

[the remark this answers is not in the document]
Comment (bkoelman) [reply docs/usage/resources/relationships.md:169]: Nope, that doesn't work:

image

[the remark this answers is not in the document]
Comment (maurei) [reply docs/usage/resources/relationships.md:169]: I see what it was supposed to do now :)

**Review by maurei — APPROVED** [review]

**Pull request merged**